src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineViewModel.cs

```diff
@@ -0,0 +1,149 @@
+        private readonly CompilationAvailableTaggerEventSource _textViewEventSource;
+        private readonly CancellationTokenSource _cancellationTokenSource;
+        private readonly IDocumentNavigationService _navigationService;
+        private readonly Workspace _workspace;
+        private CancellationToken CancellationToken => _cancellationTokenSource.Token;
+
+        public DocumentOutlineViewModel(
+            ILanguageServiceBroker2 languageServiceBroker,
+            IAsynchronousOperationListener asyncListener,
+            VisualStudioCodeWindowInfoService visualStudioCodeWindowInfoService,
+            CompilationAvailableTaggerEventSource textViewEventSource,
+            Workspace workspace,
+            IDocumentNavigationService documentNavigationService)
+        {
+            _languageServiceBroker = languageServiceBroker;
+            _visualStudioCodeWindowInfoService = visualStudioCodeWindowInfoService;
+            _textViewEventSource = textViewEventSource;
+            _workspace = workspace;
+            _navigationService = documentNavigationService;
+            _cancellationTokenSource = new CancellationTokenSource();
+
+            // initialize public properties
+            _sortOption = SortOption.Location;
+            _documentSymbolViewModelItems = new ObservableCollection<DocumentSymbolDataViewModel>();
+
+            // event queues for updating view model state
+            _expandCollapseQueue = new AsyncBatchingWorkQueue<ExpansionOption>(
+                DelayTimeSpan.NearImmediate,
+                ExpandOrCollapseItemsAsync,
+                asyncListener,
+                CancellationToken);
+
+            _selectTreeNodeQueue = new AsyncBatchingWorkQueue<CaretPosition>(
+                DelayTimeSpan.Short,
+                SelectTreeNodeAsync,
+                asyncListener,
+                CancellationToken);
+
+            _navigationQueue = new AsyncBatchingWorkQueue<TextSpan>(
+                DelayTimeSpan.NearImmediate,
+                NavigateToTextSpanAsync,
+                asyncListener,
+                CancellationToken);
+
+            _filterQueue = new AsyncBatchingWorkQueue<string>(
+                DelayTimeSpan.Short,
+                FilterTreeAsync,
+                asyncListener,
+                CancellationToken);
+
+            // work queues for refreshing LSP data
+            _documentSymbolRequestInfoQueue = new AsyncBatchingResultQueue<DocumentSymbolRequestInfo?>(
+                DelayTimeSpan.Short,
+                GetDocumentSymbolRequestInfoAsync,
+                asyncListener,
+                CancellationToken);
+
+            _documentSymbolQueue = new AsyncBatchingWorkQueue<DocumentSymbolRequestInfo, DocumentSymbolDataModel?>(
+                DelayTimeSpan.Short,
+                GetDocumentSymbolAsync,
+                EqualityComparer<DocumentSymbolRequestInfo>.Default,
+                asyncListener,
+                CancellationToken);
```

Comment:
given all the queues, it would be good to have something documented somewhere that indicates what htey are all for and how they flow through each other.

Reply:
I'll work on some ASCII art for this